src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentActivator.cs

```diff
@@ -18,49 +18,68 @@ namespace Microsoft.AspNet.Mvc.ViewComponents
     /// </remarks>
     public class DefaultViewComponentActivator : IViewComponentActivator
     {
-        private readonly Func<Type, PropertyActivator<ViewComponentContext>[]> _getPropertiesToActivate;
-        private readonly ConcurrentDictionary<Type, PropertyActivator<ViewComponentContext>[]> _injectActions;
+        private readonly ITypeActivatorCache _typeActivatorCache;
 
         /// <summary>
         /// Initializes a new instance of <see cref="DefaultViewComponentActivator"/> class.
         /// </summary>
-        public DefaultViewComponentActivator()
+        public DefaultViewComponentActivator(ITypeActivatorCache typeActivatorCache)
         {
-            _injectActions = new ConcurrentDictionary<Type, PropertyActivator<ViewComponentContext>[]>();
-            _getPropertiesToActivate = type =>
-                PropertyActivator<ViewComponentContext>.GetPropertiesToActivate(
-                    type,
-                    typeof(ViewComponentContextAttribute),
-                    CreateActivateInfo);
+            if (typeActivatorCache == null)
+            {
+                throw new ArgumentNullException(nameof(typeActivatorCache));
+            }
+
+            _typeActivatorCache = typeActivatorCache;
         }
 
         /// <inheritdoc />
-        public virtual void Activate(object viewComponent, ViewComponentContext context)
+        public virtual object Create(ViewComponentContext context)
         {
-            if (viewComponent == null)
-            {
-                throw new ArgumentNullException(nameof(viewComponent));
-            }
-
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            var propertiesToActivate = _injectActions.GetOrAdd(
-                viewComponent.GetType(),
-                _getPropertiesToActivate);
+            var componentType = context.ViewComponentDescriptor.Type.GetTypeInfo();
 
-            for (var i = 0; i < propertiesToActivate.Length; i++)
+            if (!componentType.IsClass ||
+                !componentType.IsPublic ||
```

Comment:
> We don't discover public types by default. 

This should stay. 

> But there's nothing stopping it from working in the factory.

If you use extensibility to tell us to use some random private class as a VC, then it should be supported by the factory.


> If we change it here, we might want to change it there.

We should not change `ViewComponentConventions`. What I'm saying is that the factory **could** instantiate any instantiable type. We shouldn't put a limitation in the factory.